libcst/metadata/tests/test_scope_provider.py

```diff
@@ -1982,3 +1984,210 @@ def something():
+                    class Nested:
+                        pass
+                    
+                    type Alias = Nested
+
+                    def meth1[T: Nested](self): pass
+                    def meth2[T](self, arg: Nested): pass
+            """
+        )
+        outer = ensure_type(m.body[0], cst.ClassDef)
+        nested = ensure_type(outer.body.body[0], cst.ClassDef)
+        alias = ensure_type(
+            ensure_type(outer.body.body[1], cst.SimpleStatementLine).body[0],
+            cst.TypeAlias,
+        )
+        self.assertIsInstance(scopes[alias.value], AnnotationScope)
+        nested_refs_within_alias = list(scopes[alias.value].accesses["Nested"])
+        self.assertEqual(len(nested_refs_within_alias), 1)
+        self.assertEqual(
+            {
+                ensure_type(ref, Assignment).node
+                for ref in nested_refs_within_alias[0].referents
+            },
+            {nested},
+        )
+
+        meth1 = ensure_type(outer.body.body[2], cst.FunctionDef)
+        self.assertIsInstance(scopes[meth1], ClassScope)
+        assert meth1.type_parameters
+        meth1_typevar = ensure_type(meth1.type_parameters.params[0].param, cst.TypeVar)
+        meth1_typevar_scope = scopes[meth1_typevar]
+        self.assertIsInstance(meth1_typevar_scope, AnnotationScope)
+        nested_refs_within_meth1 = list(meth1_typevar_scope.accesses["Nested"])
+        self.assertEqual(len(nested_refs_within_meth1), 1)
+        self.assertEqual(
+            {
+                ensure_type(ref, Assignment).node
+                for ref in nested_refs_within_meth1[0].referents
+            },
+            {nested},
+        )
+
+        meth2 = ensure_type(outer.body.body[3], cst.FunctionDef)
+        meth2_annotation = meth2.params.params[1].annotation
+        assert meth2_annotation
+        nested_refs_within_meth2 = list(scopes[meth2_annotation].accesses["Nested"])
+        self.assertEqual(len(nested_refs_within_meth2), 1)
+        self.assertEqual(
+            {
+                ensure_type(ref, Assignment).node
+                for ref in nested_refs_within_meth2[0].referents
+            },
+            {nested},
+        )
+
+    def test_body_isnt_subject_to_special_annotation_rule(self) -> None:
+        m, scopes = get_scope_metadata_provider(
+            """
+            class Outer:
+                class Inner: pass
+                def f[T: Inner](self): Inner()
+            """
+        )
+        outer = ensure_type(m.body[0], cst.ClassDef)
+        inner = ensure_type(outer.body.body[0], cst.ClassDef)
+        # note: this is different from global scope
+        outer_scope = scopes[outer.body.body[0]]
+        inner_assignment = list(outer_scope["Inner"])[0]
+        self.skipTest(
+            "f's body is subject to the annotation scoping rules"
+            " created by its type parameters, but it shouldn't be"
+        )
+        self.assertEqual(len(inner_assignment.references), 1)
```

Comment:
Gotta fix this somehow. The things defined in the `AnnotationScope` should still be accessible from the body, but not the immediately enclosing class assignments